pkg/nfd-master/nfd-master.go

```diff
@@ -0,0 +1,337 @@
+limitations under the License.
+*/
+
+package nfdmaster
+
+import (
+	"crypto/tls"
+	"crypto/x509"
+	"fmt"
+	"io/ioutil"
+	"log"
+	"net"
+	"os"
+	"regexp"
+	"sort"
+	"strings"
+	"time"
+
+	"golang.org/x/net/context"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials"
+	"google.golang.org/grpc/peer"
+	api "k8s.io/api/core/v1"
+	"sigs.k8s.io/node-feature-discovery/pkg/apihelper"
+	pb "sigs.k8s.io/node-feature-discovery/pkg/labeler"
+	"sigs.k8s.io/node-feature-discovery/pkg/version"
+)
+
+const (
+	// Namespace for feature labels
+	labelNs = "feature.node.kubernetes.io/"
+
+	// Namespace for all NFD-related annotations
+	annotationNs = "nfd.node.kubernetes.io/"
+)
+
+// package loggers
+var (
+	stdoutLogger = log.New(os.Stdout, "", log.LstdFlags)
+	stderrLogger = log.New(os.Stderr, "", log.LstdFlags)
+	nodeName     = os.Getenv("NODE_NAME")
+)
+
+// Labels are a Kubernetes representation of discovered features.
+type Labels map[string]string
+
+// Annotations are used for NFD-related node metadata
+type Annotations map[string]string
+
+// Command line arguments
+type Args struct {
+	CaFile         string
+	CertFile       string
+	KeyFile        string
+	LabelWhiteList *regexp.Regexp
+	NoPublish      bool
+	Port           int
+	VerifyNodeName bool
+}
+
+type NfdMaster interface {
+	Run() error
+	Stop()
+	WaitForReady(time.Duration) bool
+}
+
+type nfdMaster struct {
+	args   Args
+	server *grpc.Server
+	ready  chan bool
+}
+
+// Create new NfdMaster server instance.
+func NewNfdMaster(args Args) (*nfdMaster, error) {
+	nfd := &nfdMaster{args: args, ready: make(chan bool, 1)}
+
+	// Check TLS related args
+	if args.CertFile != "" || args.KeyFile != "" || args.CaFile != "" {
+		if args.CertFile == "" {
+			return nfd, fmt.Errorf("--cert-file needs to be specified alongside --key-file and --ca-file")
+		}
+		if args.KeyFile == "" {
+			return nfd, fmt.Errorf("--key-file needs to be specified alongside --cert-file and --ca-file")
+		}
+		if args.CaFile == "" {
+			return nfd, fmt.Errorf("--ca-file needs to be specified alongside --cert-file and --key-file")
+		}
+	}
+
+	return nfd, nil
+}
+
+// Run NfdMaster server. The method returns in case of fatal errors or if Stop()
+// is called.
+func (m *nfdMaster) Run() error {
+	stdoutLogger.Printf("Node Feature Discovery Master %s", version.Get())
+	stdoutLogger.Printf("NodeName: '%s'", nodeName)
+
+	// Initialize Kubernetes API helpers
+	helper := apihelper.APIHelpers(apihelper.K8sHelpers{})
+
+	if !m.args.NoPublish {
+		err := updateMasterNode(helper)
+		if err != nil {
+			return fmt.Errorf("failed to update master node: %v", err)
+		}
+	}
+
+	// Create server listening for TCP connections
+	lis, err := net.Listen("tcp", fmt.Sprintf(":%d", m.args.Port))
+	if err != nil {
+		return fmt.Errorf("failed to listen: %v", err)
+	}
+	// Notify that we're ready to accept connections
+	m.ready <- true
+	close(m.ready)
+	time.Sleep(time.Second)
```

Comment:
Why do we need to sleep here ?

Reply:
Well spotted! To be honest, I don't know (anymore). Must be some accidental leftover from early development/testing phase of this patchset. Removed this in the new version of the PR